### DISCUSSION TITLE
ESP32 PWM-Range via script

### DISCUSSION THREAD

**Discussion by cob555** [arendst/Tasmota#17931] [Q&A]
Hy there :-)
I have a seemingly simple problem, but could not find a simple solution yet.
The project at hand requires a PWM-output with a custom PWM range and frequency (1kHz, nothing crazy).

To be more precise, I have flashed the device using the Web-Installer so I don't really know, what features are installed.
The "On-Board-PWM" from Tasmota works fine, but by default there is no way to adjust the parameters.
 
How can I enter these commands in the Berry-Console?

Any help is greatly appreciated.

**Comment by barbudor** [comment]
If you want fine control of a PWM gpio, you can use 'setoption15 0' to use PWM (0..1023) command instead of Dimmer (0..100)
'PWMFrequency' command can be used to set the frequency

In Berry you can use 'gpio.set_pwm()' see the berry doc for arguments details

**Reply by cob555** [reply]
Thank you for your quick reply, but I am a total programming-noob. I read some of the Berry documentation, but it is hard to derive from that how to solve my problem.
I don't even know, were to type 'setoption15 0' 

The slider, which comes as standard, is basically what i want but with a custom PWM-range as mentioned (to keep it most simple).

**Reply by barbudor** [reply]
The top of my answer are standard Tasmota ocmmands that should be typed in the Console (not the Berry Console)

As you did brought Berry to the discussion so I also added a pointer to the Berry documenttation but basic usage of Tasmota doesn't requiers any knowledge in programming and Berry is totally optional

What is built into a given binary is described in [BUILDS](https://tasmota.github.io/docs/BUILDS/)

Not knowing anything about what you are trying to achieve just lead to vague answers.

**Reply by cob555** [reply]
I brought up the Berry colsole, because I read it should be preferably used with the ESP32, but I'm also fine with the standard-console off course.
Changing the PWM frequency already worked. But I don't understand the command PWMRange. The help-file states:
```
1 = reset maximum PWM range to 1023
255..1023 = set maximum PWM range
```
So there is only an upper limit for the duty cycle (with 255/1023=25% as the lower limit)?

**Reply by barbudor** [reply]
PWMrange is used to set the granularity of the PWM command
PWM command is the actual command to set the duty cycle

The duty cycle will always start as 0 for 0%
But the 100% will be the value defined by PWMrange

So if you set `PWMrange 600` you will have 0% with `PWM 0` and 100% wiht `PWM 600`
if you set `PWMrange 1` youge the default wiht 0% with `PWM 0` and 100% wiht `PWM 1023` which has the finner granularity

**Reply by cob555** [reply]
Thank you for the explanation. This is not the functionallity I was looking for (misinterpretation of the term PWMRange).
I want to adjust the "Tasmota-PWM-slider", such that:
0% slider = 10% duty cycle
100% slider = 30% duty cycle
for example.

How could that be achieved?

**Reply by barbudor** [reply]
I think you can't do that in PWM mode.
You need to be in Dimmer mode (SetOption15 1)
In that case you can use the command [`DimmerRange 10,30`](https://tasmota.github.io/docs/Commands/#dimmerrange)
This will remap the Dimmer command from 0..100 to an PWM between 10% and 30%

I don't see a way to achieve that in PWM mode.

Is that for a servo control ?

**Reply by cob555** [reply]
Yeah, now we're talkin'
DimmerRange is heading in the right direction.
Oddly enough, the set values do not correspond with the actual PWM output signal.
I get:
0% slider = 4% duty cycle
100% slider = 14% duty cycle

The parameter for DimmerRange is 0-255 (seem easy: 8 bit binary number as decimal)
Therefore 10% should equal '25' and 30% should equal '77'
For 10% it is correct. But instead of 30% I get 47%
That relation is not even linear. Slider at 50% does not give the middle of the range (22% instead of 28.5%).

So clearly I am missing something. But what?

On the plus-side the frequency is in every case stable at 1kHz.

P.S.: It is not for a servo control. I need this signal as an input for a other circuit, which only accepts a certain duty cycle range.

**Reply by barbudor** [reply]
Yes, my bad, I forgot to multiply by 255

Now you are facing non-linearity because this is designed for LED, so there is a gamma correction
Try [`LedTable 0`](https://tasmota.github.io/docs/Commands/#ledtable) to disable it

Note that all the blue text are links to the doc so you can read more details about it there

**Reply by cob555** [reply]
I noticed the blue links.

With LedTable 0 and DimmerRange 25,77 I get just some other 'odd numbers' (33% and 63% for min and max value of the slider).

There is still some fully business going on. Tasmota is apparently too smart :-D

**Reply by barbudor** [reply]
Do you have only 1 PWM ?
If not, you may need `MultiPWM 1` so they are all independant (otherwise they are considered as RGB and their might be a relation between them)

If this doesn't work, then you may have to use the standard full range 0..1023 and do calculations in rules or externally

At the end, how would you receive the commands to set the PWM ratio ?

EDIT: My bad, you want to use hte slider
And in full PWM the slider is always 0..1023

Missing ideas

**Reply by cob555** [reply]
Yes, I only need one PWM channel.

As a bodge, I could just try different values for DimmerRange until I get the desired result, but there must be also a 'proper' solution for a generic PWM, I would guess.  

Who would have thought, that problem turns out not to be as simple as it seems at first glance.

**Reply by Jason2866** [reply]
yes, since you want to use the slider, by doing this pwm is always connected to "light" functions.
There is a lot of logic "behind".

**Reply by cob555** [reply]
Yes, but I thought we omitted the logic with the command "setOption15 0"
On the other hand, the help-page states only for "setOption15 1 = control with [Color](https://tasmota.github.io/docs/Commands/#color) or [Dimmer](https://tasmota.github.io/docs/Commands/#dimmer) commands (default) "
So the dimmer commands can't work properly with this setting.

But how can I get with "setOption15 0 = basic PWM control" an easy user interface then, if the slider does not work?
Up- and down-arrows with a counter for the present duty cycle or something along those lines would also be fine.

I don't even get the slider anymore, after playing around with the PWM-channel any more... :~)

**Reply by Jason2866** [reply]
Dimmer is for lights. It does work at least for me and some others ;-)
Keep in mind the Web Frontend is NOT designed for daily use. It is a fallback Admin console

**Reply by cob555** [reply]
It may be not designed for daily use, but it is very simple, and that's what I'm after :-)

**Comment by barbudor** [comment]
May be it's time to dive in Scripting-Language or Berry to control PWM directly and inject in the WebUI a slider of your own.
I've never done that myself

**Reply by cob555** [reply]
That's an option.
For some reason I only find examples for PWM using the Arduino-IDE and none for the (presumably simpler) onboard scripting capability.
